server/project/project.go

```diff
@@ -256,31 +216,41 @@ func (s *Server) Update(ctx context.Context, q *project.ProjectUpdateRequest) (*
 		return nil, err
 	}
 
-	removedDst := getRemovedDestination(oldProj, q.Project)
-	removedSrc := getRemovedSources(oldProj, q.Project)
-
-	removedDstUsed := make([]v1alpha1.ApplicationDestination, 0)
-	removedSrcUsed := make([]string, 0)
+	var srcValidatedApps []v1alpha1.Application
+	var dstValidatedApps []v1alpha1.Application
 
 	for _, a := range argo.FilterByProjects(appsList.Items, []string{q.Project.Name}) {
-		if dest, ok := removedDst[fmt.Sprintf("%s/%s", a.Spec.Destination.Server, a.Spec.Destination.Namespace)]; ok {
-			removedDstUsed = append(removedDstUsed, dest)
+		if oldProj.IsSourcePermitted(a.Spec.Source) {
+			srcValidatedApps = append(srcValidatedApps, a)
 		}
-		if _, ok := removedSrc[a.Spec.Source.RepoURL]; ok {
-			removedSrcUsed = append(removedSrcUsed, a.Spec.Source.RepoURL)
+		if oldProj.IsDestinationPermitted(a.Spec.Destination) {
+			dstValidatedApps = append(dstValidatedApps, a)
 		}
 	}
-	if len(removedDstUsed) > 0 {
-		formattedRemovedUsedList := make([]string, len(removedDstUsed))
-		for i := 0; i < len(removedDstUsed); i++ {
-			formattedRemovedUsedList[i] = fmt.Sprintf("server: %s, namespace: %s", removedDstUsed[i].Server, removedDstUsed[i].Namespace)
+
+	invalidSrcCount := 0
+	invalidDstCount := 0
+
+	for _, a := range srcValidatedApps {
+		if !q.Project.IsSourcePermitted(a.Spec.Source) {
+			invalidSrcCount++
 		}
-		return nil, status.Errorf(
-			codes.InvalidArgument, "following destinations are used by one or more application and cannot be removed: %s", strings.Join(formattedRemovedUsedList, ";"))
 	}
-	if len(removedSrcUsed) > 0 {
-		return nil, status.Errorf(
-			codes.InvalidArgument, "following source repos are used by one or more application and cannot be removed: %s", strings.Join(removedSrcUsed, ";"))
+	for _, a := range dstValidatedApps {
+		if !q.Project.IsDestinationPermitted(a.Spec.Destination) {
+			invalidDstCount++
+		}
+	}
+
+	if invalidSrcCount > 0 || invalidDstCount > 0 {
```

Comment:
minor, but you could declare parts, append, and then check to see if `len(parts) > 0` to throw error

Reply:
good suggestion. done